src/WebJobs.Script/Workers/Rpc/FunctionRpcService.cs

```diff
@@ -52,29 +53,39 @@ public override async Task EventStream(IAsyncStreamReader<StreamingMessage> requ
                     outboundEventSubscriptions.Add(workerId, _eventManager.OfType<OutboundEvent>()
                         .Where(evt => evt.WorkerId == workerId)
                         .ObserveOn(NewThreadScheduler.Default)
-                        .Subscribe(evt =>
+                        .Subscribe(async evt =>
                         {
                             try
                             {
-                                // WriteAsync only allows one pending write at a time
-                                // For each responseStream subscription, observe as a blocking write, in series, on a new thread
                                 if (evt.MessageType == MsgType.InvocationRequest)
                                 {
-                                    _logger.LogDebug("Writing invocation request invocationId: {invocationId} to workerId: {workerId}", evt.Message.InvocationRequest.InvocationId, workerId);
+                                    _logger.LogTrace("Writing invocation request invocationId: {invocationId} to workerId: {workerId}", evt.Message.InvocationRequest.InvocationId, workerId);
+                                }
+
+                                try
+                                {
+                                    // WriteAsync only allows one pending write at a time, so we
+                                    // serialize access to the stream for each subscription
+                                    await _writeLock.WaitAsync();
+                                    await responseStream.WriteAsync(evt.Message);
+                                }
+                                finally
+                                {
+                                    _writeLock.Release();
                                 }
-                                responseStream.WriteAsync(evt.Message).GetAwaiter().GetResult();
```

Comment:
This was one of the main OOP bottlenecks. By making this async, we don't hold the thread unnecessarily anymore.

Reply:
I had attempted to make this async a while back but did not see any changes in the throughput. Assuming, this change with rest of the out-of-proc updates made a difference. Thanks!